File: examples/typescript/test/e2e/flow-smoke.test.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert/strict";
import { createServer } from "node:net";
import test from "node:test";

import { loadEnv } from "../../src/config/env.js";
import { startSampleServer, type SampleServer } from "../../src/main.js";
import {
  assertFlowSmokeNoUnexpectedFailures,
  assertFlowSmokeStartStep,
  defaultFlags,
  stepStartMayFailFlags,
  noStartStepFlags,
  employerOptInFlowId,
  type FlowSmokeContext,
  type FlowSmokeEntry,
  newFlowId,
  shortlistFlowId,
  triggerGet,
  triggerPost,
} from "./flow-smoke-helper.js";

let server: SampleServer;
let context: FlowSmokeContext;

test.before(async () => {
  process.env.DEX_EXAMPLES_HTTP_ADDRESS =
    process.env.DEX_SMOKE_HTTP_ADDRESS ?? `127.0.0.1:${await availablePort()}`;
  process.env.DEX_WORKER_BIND_ADDRESS =
    process.env.DEX_SMOKE_WORKER_BIND_ADDRESS ?? `127.0.0.1:${await availablePort()}`;
  server = await startSampleServer();
  const httpAddress = loadEnv().httpAddress;
  context = {
    baseUrl: `http://${httpAddress}`,
    newFlowId,
  };
});

test.after(async () => {
  await server.close();
});

function flowSmokeCatalog(): FlowSmokeEntry[] {
  return [
    { name: "products/engagement", trigger: () => triggerGet(context, "/products/engagement/start"), flags: defaultFlags() },
    {
      name: "products/microservices",
      trigger: () =>
        triggerGet(context, "/products/microservices/start", {
          workflowId: newFlowId("microservices"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "products/money-transfer",
      trigger: () =>
        triggerGet(context, "/products/money-transfer/start", {
          amount: 100,
          fromAccount: "from-smoke",
          toAccount: "to-smoke",
          notes: "smoke",
        }),
      flags: defaultFlags(),
    },
    {
      name: "products/order-processing",
      trigger: () => triggerGet(context, "/products/order-processing/start"),
      flags: defaultFlags(),
    },
    {
      name: "products/polling",
      trigger: () =>
        triggerGet(context, "/products/polling/start", {
          workflowId: newFlowId("product-polling"),
          pollingCompletionThreshold: 3,
        }),
      flags: defaultFlags(),
    },
    {
      name: "products/subscription",
      trigger: () => triggerGet(context, "/products/subscription/start"),
      flags: defaultFlags(),
    },
    {
      name: "products/signup",
      trigger: async () => {
        const username = newFlowId("signup");
        const result = await triggerGet(context, "/products/signup/submit", {
          username,
          email: `${username}@example.com`,
        });
        return {
          flowId: result.flowId || username,
          runId: result.runId,
        };
      },
      flags: defaultFlags(),
    },
    {
      name: "products/job-post",
      trigger: () =>
        triggerGet(context, "/products/job-post/create", {
          title: "Smoke Test Job",
          description: "Smoke test description",
        }),
      flags: noStartStepFlags(),
    },
    {
      name: "products/shortlist-candidates/employer-opt-in",
      trigger: async () => {
        const employerId = newFlowId("employer");
        await triggerPost(context, "/products/shortlist-candidates/opt_in", {
          employerId,
        });
        return { flowId: employerOptInFlowId(employerId), runId: "" };
      },
      flags: defaultFlags(),
    },
    {
      name: "products/shortlist-candidates/shortlist",
      trigger: async () => {
        const employerId = newFlowId("shortlist-employer");
        const candidateId = newFlowId("candidate");
        await triggerPost(context, "/products/shortlist-candidates/opt_in", { employerId });
        await triggerPost(context, "/products/shortlist-candidates/shortlist", {
          employerId,
          candidateId,
        });
        return { flowId: shortlistFlowId(employerId, candidateId), runId: "" };
      },
      flags: defaultFlags(),
    },
    {
      name: "patterns/polling/timer",
      trigger: () =>
        triggerGet(context, "/patterns/polling/start/timer", {
          workflowId: newFlowId("pattern-polling-simple"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/polling/backoff",
      trigger: () =>
        triggerGet(context, "/patterns/polling/start/backoff", {
          workflowId: newFlowId("pattern-polling-backoff"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/interruptible",
      trigger: () =>
        triggerGet(context, "/patterns/interruptible/start", {
          workflowId: newFlowId("interruptible"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/reminders",
      trigger: () => triggerGet(context, "/patterns/reminders/start"),
      flags: defaultFlags(),
    },
    {
      name: "patterns/entity-store",
      trigger: async () => {
        const userId = newFlowId("entity-store");
        const result = await triggerPost(context, "/patterns/entity-store/profile", {
          userId,
          displayName: "Smoke Tester",
          email: `${userId}@example.com`,
          marketingOptIn: true,
          credits: 120,
          weight: 59.5,
          lastLoggedInTime: "2026-08-11T15:30:00Z",
          metadata: { source: "smoke", tags: ["example"] },
        });
        return { flowId: result.flowId || userId, runId: result.runId };
      },
      flags: noStartStepFlags(),
    },
    {
      name: "patterns/manual-recovery",
      trigger: () =>
        triggerGet(context, "/patterns/manual-recovery/start", {
          workflowId: newFlowId("manual-recovery"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/resettable-timer",
      trigger: () =>
        triggerGet(context, "/patterns/resettable-timer/start", {
          workflowId: newFlowId("resettable-timer"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/parallel/static",
      trigger: () =>
        triggerGet(context, "/patterns/parallel/start/static", {
          workflowId: newFlowId("parallel-static"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/parallel/dynamic",
      trigger: () =>
        triggerGet(context, "/patterns/parallel/start/dynamic", {
          workflowId: newFlowId("parallel-dynamic"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/parallel/await",
      trigger: () =>
        triggerGet(context, "/patterns/parallel/start/await", {
          workflowId: newFlowId("parallel-await"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/parallel/first-win",
      trigger: () =>
        triggerGet(context, "/patterns/parallel/start/first-win", {
          workflowId: newFlowId("parallel-first-win"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/recovery",
      trigger: () =>
        triggerGet(context, "/patterns/recovery/start", {
          workflowId: newFlowId("recovery"),
          itemName: "smoke-item",
          quantity: 2,
        }),
      flags: stepStartMayFailFlags(),
    },
    {
      name: "patterns/scalable-parallel",
      trigger: () =>
        triggerGet(context, "/patterns/scalable-parallel/start", {
          workflowId: newFlowId("scalable-parallel"),
          numOfChildWfs: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/parent-child",
      trigger: () =>
        triggerGet(context, "/patterns/parent-child/start", {
          workflowId: newFlowId("parent-child"),
          numOfChildWfs: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/drain-channels/internal",
      trigger: () =>
        triggerGet(context, "/patterns/drain-channels/internal/start", {
          workflowId: newFlowId("drain-internal"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/drain-channels/external-publishing",
      trigger: () =>
        triggerGet(context, "/patterns/drain-channels/external-publishing/start-or-publish", {
          workflowId: newFlowId("drain-external"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/wait-for-state-completion",
      trigger: () =>
        triggerGet(context, "/patterns/wait-for-state-completion/start", {
          workflowId: newFlowId("wait-for-state"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "patterns/timeout",
      trigger: () =>
        triggerGet(context, "/patterns/timeout/start", {
          workflowId: newFlowId("timeout"),
          successfulWorkflow: true,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step",
      trigger: () =>
        triggerGet(context, "/primitives/step/start", {
          workflowId: newFlowId("primitive-step"),
          inputNum: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/retry",
      trigger: () =>
        triggerGet(context, "/primitives/step/retry/start", {
          workflowId: newFlowId("primitive-step-retry"),
          readyAfterAttempt: 2,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/custom-retry",
      trigger: () =>
        triggerGet(context, "/primitives/step/custom-retry/start", {
          workflowId: newFlowId("primitive-step-custom-retry"),
          readyAfterAttempt: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/durability",
      trigger: () =>
        triggerGet(context, "/primitives/step/durability/start", {
          workflowId: newFlowId("primitive-step-durability"),
          mode: "sync",
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/heartbeat",
      trigger: () =>
        triggerGet(context, "/primitives/step/heartbeat/start", {
          workflowId: newFlowId("primitive-step-heartbeat"),
          batches: 0,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/options-override",
      trigger: () =>
        triggerGet(context, "/primitives/step/options-override/start", {
          workflowId: newFlowId("primitive-step-options-override"),
          input: "smoke",
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/step-decision",
      trigger: () =>
        triggerGet(context, "/primitives/step/step-decision/start", {
          workflowId: newFlowId("primitive-step-decision"),
          mode: "graceful",
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/step/wait-types",
      trigger: () =>
        triggerGet(context, "/primitives/step/wait-types/start", {
          workflowId: newFlowId("primitive-step-wait-types"),
          mode: "any",
          timeoutSeconds: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/attribute",
      trigger: () =>
        triggerGet(context, "/primitives/attribute/start", {
          workflowId: newFlowId("primitive-attribute"),
          message: "smoke",
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/channel",
      trigger: () =>
        triggerGet(context, "/primitives/channel/start", {
          workflowId: newFlowId("primitive-channel"),
          inputNum: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/stream",
      trigger: () =>
        triggerGet(context, "/primitives/stream/start", {
          workflowId: newFlowId("primitive-stream"),
          input: "smoke",
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/timer",
      trigger: () =>
        triggerGet(context, "/primitives/timer/start", {
          workflowId: newFlowId("primitive-timer"),
          seconds: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/rpc",
      trigger: () =>
        triggerGet(context, "/primitives/rpc/start", {
          workflowId: newFlowId("primitive-rpc"),
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/subflow",
      trigger: () =>
        triggerGet(context, "/primitives/subflow/start", {
          workflowId: newFlowId("primitive-subflow"),
          inputNum: 1,
        }),
      flags: defaultFlags(),
    },
    {
      name: "primitives/client-apis",
      trigger: () =>
        triggerGet(context, "/primitives/client-apis/start", {
          workflowId: newFlowId("primitive-client-apis"),
          keyword: "smoke",
        }),
      flags: defaultFlags(),
    },
  ];
}

test("flow smoke catalog size", () => {
  const catalog = flowSmokeCatalog();
  assert.ok(catalog.length > 0, "flow smoke catalog is empty");
});

test("flow smoke all registered flows via controller", async () => {
  for (const entry of flowSmokeCatalog()) {
    try {
      const result = await entry.trigger(context);
      assert.ok(result.flowId, `${entry.name}: controller response did not include flowID`);
      await assertFlowSmokeStartStep(entry, result.flowId, result.runId);
      if (entry.name === "patterns/interruptible") {
        await triggerGet(context, "/patterns/interruptible/cancel", {
          workflowId: result.flowId,
        });
      }
      await assertFlowSmokeNoUnexpectedFailures(entry, result.flowId, result.runId);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`${entry.name}: ${message}`, { cause: error });
    }
  }
});

function availablePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const server = createServer();
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      if (address === null || typeof address === "string") {
        server.close();
        reject(new Error("failed to allocate a smoke test port"));
        return;
      }
      server.close((error) => {
        if (error !== undefined) {
          reject(error);
          return;
        }
        resolve(address.port);
      });
    });
  });
}

File: examples/go/integ/flow_smoke_test.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package integ

import (
	"net/http"
	"net/url"
	"testing"
	"time"

	"github.com/superdurable/dex/examples/go/patterns/entity-store"
)

func flowSmokeCatalog() []flowSmokeEntry {
	return []flowSmokeEntry{
		{
			name: "products/engagement",
			trigger: func(t *testing.T) (string, string) {
				flowID, runID := triggerFlowSmokeHTTP(t, http.MethodGet, "/products/engagement/start", nil, nil)
				return flowID, runID
			},
		},
		{
			name: "products/microservices",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "microservices")}}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/products/microservices/start", query, nil)
			},
		},
		{
			name: "products/money-transfer",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"amount":      {"100"},
					"fromAccount": {"from-smoke"},
					"toAccount":   {"to-smoke"},
					"notes":       {"smoke"},
				}
				flowID, runID := triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/products/money-transfer/start",
					query,
					nil,
				)
				return flowID, runID
			},
		},
		{
			name: "products/order-processing",
			trigger: func(t *testing.T) (string, string) {
				flowID, runID := triggerFlowSmokeHTTP(t, http.MethodGet, "/products/order-processing/start", nil, nil)
				return flowID, runID
			},
		},
		{
			name: "products/subscription",
			trigger: func(t *testing.T) (string, string) {
				flowID, runID := triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/products/subscription/start",
					nil,
					nil,
				)
				return flowID, runID
			},
		},
		{
			name: "products/signup",
			trigger: func(t *testing.T) (string, string) {
				username := smokeWorkflowID(t, "signup")
				query := url.Values{
					"username": {username},
					"email":    {username + "@example.com"},
				}
				flowID, runID := triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/products/signup/submit",
					query,
					nil,
				)
				return flowID, runID
			},
		},
		{
			name: "products/job-post",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"title":       {"Smoke Test Job"},
					"description": {"Smoke test description"},
				}
				flowID, runID := triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/products/job-post/create",
					query,
					nil,
				)
				return flowID, runID
			},
			flags: flowSmokeFlags{noStartStep: true},
		},
		{
			name: "patterns/polling/timer",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "pattern-polling-simple")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/polling/start/timer",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/polling/backoff",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "pattern-polling-backoff")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/polling/start/backoff",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/polling/iteration",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "pattern-polling-iteration")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/polling/start/iteration",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/interruptible",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "interruptible")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/interruptible/start",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/reminders",
			trigger: func(t *testing.T) (string, string) {
				_, _, responseBody := triggerFlowSmokeHTTPWithBody(
					t,
					http.MethodGet,
					"/patterns/reminders/start",
					nil,
					nil,
				)
				flowID, runID := parseFlowTriggerResponse(string(responseBody), "")
				return flowID, runID
			},
		},
		{
			name: "patterns/entity-store",
			trigger: func(t *testing.T) (string, string) {
				userID := smokeWorkflowID(t, "entity-store")
				body := entitystore.UserProfileRequest{
					UserID: userID,
					UserProfile: entitystore.UserProfile{
						DisplayName:    "Smoke Tester",
						Email:          userID + "@example.com",
						MarketingOptIn: true,
						Credits:        120,
						Weight:         59.5,
						LastLoggedIn:   time.Date(2026, 8, 11, 15, 30, 0, 0, time.UTC),
						Metadata: entitystore.UserProfileMetadata{
							Source: "smoke",
							Tags:   []string{"example"},
						},
					},
				}
				flowID, runID := triggerFlowSmokeHTTP(
					t,
					http.MethodPost,
					"/patterns/entity-store/profile",
					nil,
					body,
				)
				if flowID == "" {
					flowID = userID
				}
				return flowID, runID
			},
			flags: flowSmokeFlags{noStartStep: true},
		},
		{
			name: "patterns/manual-recovery",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "manual-recovery")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/manual-recovery/start",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/inactiveness-tracker-timer",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "inactiveness-tracker-timer")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/inactiveness-tracker-timer/start",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/parallel/static",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "parallel-static")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/parallel/start/static",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/parallel/dynamic",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "parallel-dynamic")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/parallel/start/dynamic",
					query,
					nil,
				)
			},
		},
		parallelSmokeEntry("await"),
		parallelSmokeEntry("first-win"),
		{
			name: "patterns/recovery",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "recovery")},
					"itemName":   {"smoke-item"},
					"quantity":   {"2"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/patterns/recovery/start", query, nil)
			},
			flags: flowSmokeFlags{stepStartMayFail: true},
		},
		parallelSubFlowsSmokeEntry("basic"),
		parallelSubFlowsSmokeEntry("wait-for-half"),
		parallelSubFlowsSmokeEntry("long-lived-parent"),
		parallelSubFlowsSmokeEntry("short-lived-parent"),
		{
			name: "patterns/drain-channels/internal",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "drain-internal")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/drain-channels/internal/start",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/drain-channels/external-publishing",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "drain-external")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/drain-channels/external-publishing/start-or-publish",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/wait-for-step-completion",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "wait-for-state")}}
				return triggerFlowSmokeHTTP(
					t,
					http.MethodGet,
					"/patterns/wait-for-step-completion/start",
					query,
					nil,
				)
			},
		},
		{
			name: "patterns/timeout",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId":         {smokeWorkflowID(t, "timeout")},
					"successfulWorkflow": {"true"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/patterns/timeout/start", query, nil)
			},
		},
		{
			name: "primitives/step",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-step")},
					"inputNum":   {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/start", query, nil)
			},
		},
		{
			name: "primitives/step/retry",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId":        {smokeWorkflowID(t, "primitive-step-retry")},
					"readyAfterAttempt": {"2"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/retry/start", query, nil)
			},
		},
		{
			name: "primitives/step/custom-retry",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId":        {smokeWorkflowID(t, "primitive-step-custom-retry")},
					"readyAfterAttempt": {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/custom-retry/start", query, nil)
			},
		},
		{
			name: "primitives/step/durability",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-step-durability")},
					"mode":       {"sync"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/durability/start", query, nil)
			},
		},
		{
			name: "primitives/step/heartbeat",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-step-heartbeat")},
					"batches":    {"0"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/heartbeat/start", query, nil)
			},
		},
		{
			name: "primitives/step/options-override",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-step-options-override")},
					"input":      {"smoke"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/options-override/start", query, nil)
			},
		},
		{
			name: "primitives/step/step-decision",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-step-decision")},
					"mode":       {"graceful"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/step-decision/start", query, nil)
			},
		},
		{
			name: "primitives/step/wait-types",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId":     {smokeWorkflowID(t, "primitive-step-wait-types")},
					"mode":           {"any"},
					"timeoutSeconds": {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/step/wait-types/start", query, nil)
			},
		},
		{
			name: "primitives/attribute",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-attribute")},
					"message":    {"smoke"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/attribute/start", query, nil)
			},
		},
		{
			name: "primitives/channel",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-channel")},
					"inputNum":   {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/channel/start", query, nil)
			},
		},
		{
			name: "primitives/stream",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-stream")},
					"input":      {"smoke"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/stream/start", query, nil)
			},
		},
		{
			name: "primitives/timer",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-timer")},
					"seconds":    {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/timer/start", query, nil)
			},
		},
		{
			name: "primitives/rpc",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{"workflowId": {smokeWorkflowID(t, "primitive-rpc")}}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/rpc/start", query, nil)
			},
		},
		{
			name: "primitives/subflow",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-subflow")},
					"inputNum":   {"1"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/subflow/start", query, nil)
			},
		},
		{
			name: "primitives/client-apis",
			trigger: func(t *testing.T) (string, string) {
				query := url.Values{
					"workflowId": {smokeWorkflowID(t, "primitive-client-apis")},
					"keyword":    {"smoke"},
				}
				return triggerFlowSmokeHTTP(t, http.MethodGet, "/primitives/client-apis/start", query, nil)
			},
		},
	}
}

func parallelSmokeEntry(kind string) flowSmokeEntry {
	return flowSmokeEntry{
		name: "patterns/parallel/" + kind,
		trigger: func(t *testing.T) (string, string) {
			query := url.Values{"workflowId": {smokeWorkflowID(t, "parallel-"+kind)}}
			return triggerFlowSmokeHTTP(
				t,
				http.MethodGet,
				"/patterns/parallel/start/"+kind,
				query,
				nil,
			)
		},
	}
}

func parallelSubFlowsSmokeEntry(kind string) flowSmokeEntry {
	return flowSmokeEntry{
		name: "patterns/parallel-subflows/" + kind,
		trigger: func(t *testing.T) (string, string) {
			query := url.Values{"workflowId": {smokeWorkflowID(t, "parallel-subflows-"+kind)}}
			return triggerFlowSmokeHTTP(
				t,
				http.MethodGet,
				"/patterns/parallel-subflows/start/"+kind,
				query,
				nil,
			)
		},
	}
}

func TestFlowSmokeCatalogSize(t *testing.T) {
	catalog := flowSmokeCatalog()
	if len(catalog) == 0 {
		t.Fatal("flow smoke catalog is empty")
	}
	t.Logf("flow smoke catalog size: %d", len(catalog))
}

func TestFlowSmokeAllRegisteredFlowsViaController(t *testing.T) {
	for _, entry := range flowSmokeCatalog() {
		entry := entry
		t.Run(entry.name, func(t *testing.T) {
			flowID, runID := entry.trigger(t)
			requireFlowSmokeIDs(t, entry.name, flowID)
			assertFlowSmokeStartStep(t, entry, flowID, runID)
			assertFlowSmokeNoUnexpectedFailures(t, entry, flowID, runID)
		})
	}
}

func requireFlowSmokeIDs(t *testing.T, name string, flowID string) {
	t.Helper()
	if flowID == "" {
		t.Fatalf("%s: controller response did not include flowID", name)
	}
}
